enterprise/views.py

```diff
@@ -952,3 +964,376 @@ def get(self, request, enterprise_uuid, course_id):
+        related information.
+
+        :param username: The learner who might be enrolled in the program.
+        :param program_details: The details of the program to check enrollment against.
+        :return: Whether the learner is enrolled in the program and whether they're program certificate-eligible.
+        """
+        # A single violation in the logic relevant to these will flip their booleans and keep them that way thereon.
+        enrolled_in_program = False
+        certificate_eligible = True
+
+        program_courses = []
+        enrollment_client = EnrollmentApiClient()
+        for program_course in program_details['courses']:
+            # Safely make the assumption that the first course run is available and is the one we want.
+            program_course_run = program_course['course_runs'][0]
+            course_id = program_course_run['key']
+            # Track courses so we can calculate the price later.
+            course_modes = enrollment_client.get_course_modes(course_id)
+            course_mode_for_calculation = [mode for mode in course_modes if mode['slug'] == 'verified'][0]
+
+            enrolled_course = enrollment_client.get_course_enrollment(user.username, course_id)
+            paid_for_course_certificate = bool(
+                enrolled_course and
+                enrolled_course['mode'] in getattr(
+                    settings,
+                    'ENTERPRISE_COURSE_ENROLLMENT_VERIFIED_MODES',
+                    ['verified', 'professional']
+                )
+            )
+            if paid_for_course_certificate:  # pylint: disable=simplifiable-if-statement
+                # Abide by the definition of program enrollment, i.e. enrollment in >= 1 course == program enrollment.
+                # Do not simplify this if-statement to `enrolled_in_program = paid_for_course_certificate`,
+                # because these are just one-time switches.
+                enrolled_in_program = True
+            else:
+                # A single course whose certificate is unpaid for == not program certificate-eligible.
+                certificate_eligible = False
+
+            program_course.update({
+                'course_id': course_id,
+                'enrolled': enrolled_course is not None,
+                'paid_for_certificate': paid_for_course_certificate,
+                'course_mode': course_mode_for_calculation,
+            })
+            program_courses.append(program_course)
+
+        program_price, program_discounted_price = self.get_program_prices(program_courses, user)
+        return {
+            'enrolled_in_program': enrolled_in_program,
+            'certificate_eligible_for_program': certificate_eligible,
+            'program_price': program_price,
+            'program_discounted_price': program_discounted_price,
+            'courses': program_courses,
+        }
```

Comment:
@UmanShahzad There is prior art for calculating the learner-specific program information we need to render the program enrollment page.

https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/programs/utils.py#L454

We may want to consider refactoring some of this existing code to make it usable by our enterprise code. We may also want to consider breaking our standard approach of using REST APIs for communicating with LMS code for performance sake.

Reply:
@douglashall Just to confirm -- are you asking to import these utilities from the platform and re-use them here in place of the current changeset, or to refactor this changeset to be more like the utilities in the platform, and to be reusable throughout the edx-enterprise repo?

As for not using REST APIs, I'm wondering what else you have in mind to replace that -- in many cases, we simply need a lot of the information that the endpoint provides.

Reply:
My initial thinking was to refactor the edx-platform code linked above to allow us to import a single function that would do the "one-click purchase eligibility" calculation and return it to us, but now I'm thinking we just import `ProgramDataExtender`, pass it the program data dictionary, and let it do it's thing, instead of recreating the logic here.

As for performance, I think it would be better to avoid several API roundtrips to render a single page. We could create a new function on the ecommerce API client that lives in edx-enterprise which wraps the call to `ProgramDataExtender` and returns the "extended" program data.

Reply:
Ah, I understand what you mean. So we can do that, but 2 things:

1. This would entail a lot of code changes, especially in the tests (and the test code is large), and considering it's a refactor and we're (I believe) under pressure with a deadline, we should maybe consider focusing most on testing and let the refactor happen in the follow-up story (i.e. adding details) or in a separate story from these both, whatever seems appropriate.
1. This makes me wonder why this repository isn't just a part of the platform directly if solutions are best done by just  :laughing: